tests/tb_device_mqtt_client_tests.py

```diff
@@ -82,19 +86,425 @@ def test_send_telemetry_and_attr(self):
+        if self.subscribe_to_attribute_all is not None:
+            self.assertEqual(self.subscribe_to_attribute_all, {self.shared_attribute_name: value})
+        else:
+            self.fail("subscribe_to_attribute_all is None")
+
+        if self.subscribe_to_attribute is not None:
+            self.assertEqual(self.subscribe_to_attribute, {self.shared_attribute_name: value})
+        else:
+            self.fail("subscribe_to_attribute is None")
 
         self.client.unsubscribe_from_attribute(sub_id_1)
         self.client.unsubscribe_from_attribute(sub_id_2)
 
+    def test_send_rpc_call(self):
+        def rpc_callback(req_id, result, exception):
+            self.assertEqual(result, {"response": "success"})
+            self.assertIsNone(exception)
+
+        self.client.send_rpc_call("testMethod", {"param": "value"}, rpc_callback)
+
+    def test_publish_with_error(self):
+        with self.assertRaises(TBQoSException):
+            self.client._publish_data("invalid", "invalid_topic", qos=3)
+
+    def test_decode_message(self):
+        mock_message = MagicMock()
+        mock_message.payload = b'{"key": "value"}'
+        decoded = self.client._decode(mock_message)
+        self.assertEqual(decoded, {"key": "value"})
+
+    def test_decode_message_valid_json_str(self):
+        mock_message = MagicMock()
+        mock_message.payload = '{"foo": "bar"}'
+        decoded = self.client._decode(mock_message)
+        self.assertEqual(decoded, {"foo": "bar"})
+
+    def test_decode_message_invalid_json_but_valid_utf8_str(self):
+        mock_message = MagicMock()
+        mock_message.payload = 'invalid {json:'
+        with self.assertRaises(AttributeError):
+            self.client._decode(mock_message)
+
+    def test_decode_message_invalid_json_bytes(self):
+        mock_message = MagicMock()
+        mock_message.payload = b'invalid json data'
+        decoded = self.client._decode(mock_message)
+        self.assertEqual(decoded, "invalid json data")
+
+    def test_decode_message_invalid_utf8_bytes(self):
+        mock_message = MagicMock()
+        mock_message.payload = b'\xff\xfe\xfa'
+        decoded = self.client._decode(mock_message)
+        self.assertEqual(decoded, '')
+
+    def test_on_decoded_message_rpc_request(self):
+        client = TBDeviceMqttClient(host="test_host", port=1883, username="test_token")
+        client._messages_rate_limit = MagicMock()
+        mock_rpc_handler = MagicMock()
+        client.set_server_side_rpc_request_handler(mock_rpc_handler)
+        message = MagicMock()
+        message.topic = RPC_REQUEST_TOPIC + "42"
+        message.payload = b'{"some_key": "some_value"}'
+        content = {"some_key": "some_value"}
+        client._on_decoded_message(content, message)
+        client._messages_rate_limit.increase_rate_limit_counter.assert_called_once()
+        mock_rpc_handler.assert_called_once_with("42", content)
+
+    def test_max_inflight_messages_set(self):
```

Comment:
In max_inflight_messages_set can be 2 variants of setting _max_inflight_messages, if passed value more than 0 or less than 0, we need to test the both cases.